src/safeposix/syscalls/fs_calls.rs

```diff
@@ -3058,91 +3144,180 @@ impl Cage {
+                            let inodenum = ui.inode;
+                            // Decrement the reference count for the send pipe if it exists. If the
+                            // reference count drops to zero, remove the send pipe.
+                            if let Some(sendpipe) = ui.sendpipe.as_ref() {
+                                sendpipe.decr_ref(O_WRONLY);
+                                //last reference, lets remove it
+                                if sendpipe.is_pipe_closed() {
+                                    ui.sendpipe = None;
+                                }
                             }
-                        }
-                        if let Some(receivepipe) = ui.receivepipe.as_ref() {
-                            receivepipe.decr_ref(O_RDONLY);
-                            //last reference, lets remove it
-                            if receivepipe.is_pipe_closed() {
-                                ui.receivepipe = None;
+                            // Decrement the reference count for the receive pipe if it exists. If
+                            // the reference count drops to zero, remove
+                            // the receive pipe.
+                            if let Some(receivepipe) = ui.receivepipe.as_ref() {
+                                receivepipe.decr_ref(O_RDONLY);
+                                //last reference, lets remove it
+                                if receivepipe.is_pipe_closed() {
+                                    ui.receivepipe = None;
+                                }
                             }
-                        }
-                        let mut inodeobj = FS_METADATA.inodetable.get_mut(&ui.inode).unwrap();
-                        if let Inode::Socket(ref mut sock) = *inodeobj {
-                            sock.refcount -= 1;
-                            if sock.refcount == 0 {
-                                if sock.linkcount == 0 {
-                                    drop(inodeobj);
-                                    let path = normpath(
-                                        convpath(sockhandle.localaddr.unwrap().path()),
-                                        self,
-                                    );
-                                    FS_METADATA.inodetable.remove(&inodenum);
-                                    NET_METADATA.domsock_paths.remove(&path);
+                            // Retrieve the inode object for the socket and decrement its reference
+                            // count. If both the reference count and
+                            // link count are zero, the socket is no longer needed.
+                            let mut inodeobj = FS_METADATA.inodetable.get_mut(&ui.inode).unwrap();
+                            if let Inode::Socket(ref mut sock) = *inodeobj {
+                                sock.refcount -= 1;
+                                if sock.refcount == 0 {
```

Comment:
Shouldn't we combine the two if statements with an AND operator?

Reply:
Updated.